src/frontend/src/optimizer/plan_node/stream_hash_join.rs

```diff
@@ -232,36 +242,66 @@ impl StreamNode for StreamHashJoin {
     }
 }
 
-fn infer_internal_table_catalog(input: PlanRef, join_key_indices: Vec<usize>) -> TableCatalog {
+/// Return hash join internal table catalog and degree table catalog.
+fn infer_internal_and_degree_table_catalog(
+    input: PlanRef,
+    join_key_indices: Vec<usize>,
+) -> (TableCatalog, TableCatalog) {
     let base = input.plan_base();
     let schema = &base.schema;
 
     let append_only = input.append_only();
-    let dist_keys = base.dist.dist_column_indices().to_vec();
 
-    // The pk of hash join internal table should be join_key + input_pk.
+    let internal_table_dist_keys = base.dist.dist_column_indices().to_vec();
+
+    // Find the dist key position in join key.
+    // FIXME(yuhao): currently the dist key position is not the exact position mapped to the join
+    // key when there are duplicate value in join key indices.
+    let degree_table_dist_keys = internal_table_dist_keys
+        .iter()
+        .map(|idx| {
+            join_key_indices
+                .iter()
+                .position(|v| v == idx)
+                .expect("join key should contain dist key.")
+        })
+        .collect_vec();
+
+    // The pk of hash join internal and degree table should be join_key + input_pk.
     let mut pk_indices = join_key_indices;
     // TODO(yuhao): dedup the dist key and pk.
     pk_indices.extend(&base.logical_pk);
 
-    let mut columns_fields = schema.fields().to_vec();
-
-    // The join degree at the end of internal table.
-    let degree_column_field = Field::with_name(DataType::Int64, "_degree");
-    columns_fields.push(degree_column_field);
-
+    // Build internal table
     let mut internal_table_catalog_builder = TableCatalogBuilder::new();
+    let internal_columns_fields = schema.fields().to_vec();
 
-    columns_fields.iter().for_each(|field| {
+    internal_columns_fields.iter().for_each(|field| {
         internal_table_catalog_builder.add_column(field);
     });
 
     pk_indices.iter().for_each(|idx| {
         internal_table_catalog_builder.add_order_column(*idx, OrderType::Ascending)
     });
 
+    // Build degree table.
+    let mut degree_table_catalog_builder = TableCatalogBuilder::new();
+
+    let degree_column_field = Field::with_name(DataType::Int64, "_degree");
+
+    pk_indices.iter().enumerate().for_each(|(order_idx, idx)| {
```

Comment:
I guess the read for degree is not going to be co-located, i.e., we have a potentially massive random read workload for join key with large number of rows.

Reply:
Yes, now the degree can not be co-located with data from upstream now. But since we are range scanning a join key, why a large number of rows will be even worse in locality?